ts/features/bonus/bpd/saga/orchestration/optInPaymentMethods/optInShouldShowChoiceHandler.ts

```diff
@@ -0,0 +1,104 @@
+import { put, select, take } from "typed-redux-saga/macro";
+import { ActionType, getType, isActionOf } from "typesafe-actions";
+import * as pot from "italia-ts-commons/lib/pot";
+import {
+  fetchWalletsFailure,
+  fetchWalletsRequestWithExpBackoff,
+  fetchWalletsSuccess
+} from "../../../../../../store/actions/wallet/wallets";
+import { bpdAllData } from "../../../store/actions/details";
+import {
+  optInPaymentMethodsShowChoice,
+  optInPaymentMethodsStart
+} from "../../../store/actions/optInPaymentMethods";
+import {
+  bpdEnabledSelector,
+  optInStatusSelector
+} from "../../../store/reducers/details/activation";
+import { CitizenOptInStatusEnum } from "../../../../../../../definitions/bpd/citizen_v2/CitizenOptInStatus";
+import { ReduxSagaEffect } from "../../../../../../types/utils";
+
+/**
+ * This saga manage the flow that checks if a user has already take a choice about the opt-in of the payment methods.
+ *
+ *  The saga follows this flow::
+ * - request the bpd data
+ * - check if the user participate or not in the cashback program
+ * - check if the user has already taken the opt-in payment methods choice
+ * - request the user's payment methods
+ * - check if the loading of the payment method succeed
+ * - if succeed start the saga
+ */
+export function* optInShouldShowChoiceHandler(): Generator<
+  ReduxSagaEffect,
+  void,
+  any
+> {
+  // Load the information about the participation of the user to the bpd program
+  yield* put(bpdAllData.request());
+  const bpdAllDataResponse = yield* take<
+    ActionType<typeof bpdAllData.success | typeof bpdAllData.failure>
+  >([getType(bpdAllData.success), getType(bpdAllData.failure)]);
+
+  // If the bpdAllData request fail report the error
+  if (isActionOf(bpdAllData.failure, bpdAllDataResponse)) {
+    yield* put(
+      optInPaymentMethodsShowChoice.failure(bpdAllDataResponse.payload)
+    );
+    return;
+  }
+
+  const bpdEnabled: pot.Pot<boolean, Error> = yield* select(bpdEnabledSelector);
+
+  // Safety check on field returned in @link{bpdEnabled} and managed by @{enabledReducer}
+  if (bpdEnabled.kind !== "PotSome") {
+    yield* put(
+      optInPaymentMethodsShowChoice.failure(
+        new Error("The bpdEnabled value is not potSome")
+      )
+    );
+    return;
+  }
+
+  // If the user is not enrolled in the bpd program returns with value false
+  if (!bpdEnabled.value) {
+    yield* put(optInPaymentMethodsShowChoice.success(bpdEnabled.value));
+    return;
+  }
```

Comment:
At the moment, from a store point of view, we don't distinguish between a user who has never been enrolled to cashback (response `404`) and a user who has been registered but has unsubscribed (`200` with body `enabled: false`). We map both these cases to `enabled: false`.
If I understand correctly, with the current implementation, we never ask someone who has unsubscribed from cashback if he wants to delete their payment methods. It's correct?

Reply:
You're right. I initially misunderstood the semantics of the response.
Refactored in [d44d979](https://github.com/pagopa/io-app/pull/3785/commits/d44d97909f9b232ee144ce0d984842447b60e643), [31abdcd](https://github.com/pagopa/io-app/pull/3785/commits/31abdcd8eb6c2aa6efd8d3a2779a46f997c949e3)